packages/auto-id/src/certificateManager.ts

```diff
@@ -0,0 +1,346 @@
+      issuer: csr.subject,
+      subject: csr.subject,
+      notBefore,
+      notAfter,
+      signingAlgorithm: privateKey.algorithm,
+      publicKey: publicKey,
+      signingKey: privateKey,
+    })
+
+    const autoIdSan = `autoid:auto:${Buffer.from(autoId).toString('hex')}`
+
+    const sanExtensions = csr.extensions.filter((ext) => ext.type === OID_SUBJECT_ALT_NAME) // OID for subjectAltName
+    if (sanExtensions.length) {
+      // const existingSan = sanExtensions[0].value
+      const existingSan = sanExtensions[0] as x509.SubjectAlternativeNameExtension
+
+      const generalNames = existingSan.names.toJSON()
+
+      // Add autoIdSan to generalNames
+      generalNames.push({
+        type: 'url' as x509.GeneralNameType,
+        value: autoIdSan,
+      })
+
+      // const newSanExtension = existingSan + CertificateManager.stringToArrayBuffer(autoIdSan)
+      const newSanExtension = new x509.SubjectAlternativeNameExtension(
+        generalNames,
+        existingSan.critical,
+      )
+      certificateBuilder.extensions.push(newSanExtension)
+    } else {
+      // certificateBuilder.extensions.push(
+      //   new x509.SubjectAlternativeNameExtension([autoIdSan]),
+      //   false,
+      // )
+
+      certificateBuilder.extensions.push(
+        new x509.SubjectAlternativeNameExtension([
+          { type: 'url' /* as x509.GeneralNameType */, value: autoIdSan },
+        ]),
+      )
+    }
+
+    // Copy all extensions from the CSR to the certificate
+    for (const ext of csr.extensions) {
+      // certificateBuilder.extensions.push(new x509.Extension(ext.value, ext.critical))
+      certificateBuilder.extensions.push(ext)
+    }
+
+    const certificateSigned = await x509.X509CertificateGenerator.create({
+      serialNumber: certificateBuilder.serialNumber,
+      issuer: certificateBuilder.issuer,
+      subject: certificateBuilder.subject,
+      notBefore: certificateBuilder.notBefore,
+      notAfter: certificateBuilder.notAfter,
+      extensions: certificateBuilder.extensions,
+      publicKey: certificateBuilder.publicKey,
+      signingAlgorithm: certificateBuilder.signatureAlgorithm,
+      signingKey: privateKey,
+    })
+
+    return certificateSigned
+  }
+
+  async selfIssueCertificate(
+    subjectName: string,
+    validityPeriodDays: number = 365,
+  ): Promise<x509.X509Certificate> {
+    if (!this.privateKey || !this.publicKey) {
+      throw new Error('Private or public key is not set.')
+    }
+
+    const csr = await this.createAndSignCSR(subjectName)
+    const certificate = await this.issueCertificate(csr, validityPeriodDays)
+
+    this.certificate = certificate
+    return certificate
+  }
+
+  saveCertificate(filePath: string): void {
+    if (!this.certificate) {
+      throw new Error('No certificate available to save.')
+    }
+
+    const certificatePem = CertificateManager.certificateToPem(this.certificate)
+    fs.writeFileSync(filePath, certificatePem, 'utf8')
```

Comment:
Why not use [save()](https://github.com/subspace/auto-sdk/blob/fa02528f2fdb887e30c4c96c1f20d8e34c81c546/packages/auto-utils/src/save.ts#L1) here?